### ISSUE TITLE
Freezing parameters: multi_transform + set_to_zero does not work?

### ISSUE THREAD

**Issue by AndreaCossu** [google-deepmind/optax#477]
Hi everyone!
I know you already received a bunch of questions on how to freeze parameters ( #175 #290...) and that you are also working on a new example on that (#296).
However, I am still facing some issues, even in a very basic setup. 

My objective is to freeze a subset of the parameters to prevent them from changing during training.

I started from the [Flax Quick Start code](https://flax.readthedocs.io/en/latest/getting_started.html) to train a CNN on MNIST by freezing the conv layer and training only the dense one. 
To do that, I modified the `create_train_state` function:

```python
def create_train_state(module, rng, learning_rate, momentum):
  params = module.init(rng, jnp.ones([1, 28, 28, 1]))['params'] 
  
  # I added this line to update only Dense_0
  tx = optax.multi_transform({"train": optax.sgd(learning_rate, momentum), "zero": optax.set_to_zero()}, 
                             freeze({"Dense_0": "train", "Conv_0": "zero"}))

  return train_state.TrainState.create(
      apply_fn=module.apply, params=params, tx=tx)
```

The model I am using is a very simple one

```python
class CNN(nn.Module):
  @nn.compact
  def __call__(self, x):
    x = nn.Conv(features=32, kernel_size=(3, 3))(x)  # Conv_0
    x = nn.relu(x)
    x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
    x = x.reshape((x.shape[0], -1))
    x = nn.Dense(features=10)(x)  # Dense_0
    return x
```

However, if I print gradients after a single training iteration, they are not zero for "Conv_0" and in fact all parameters get updated. Am I missing something?

Here is the full code for reference:

```python
import tensorflow_datasets as tfds 
import tensorflow as tf            
from flax import linen as nn  
import jax
import jax.numpy as jnp  
from flax.training import train_state  
from flax import struct 
from flax.core import unfreeze, freeze
import optax 

def create_train_state(module, rng, learning_rate, momentum):
  params = module.init(rng, jnp.ones([1, 28, 28, 1]))['params'] 
  tx = optax.multi_transform({"train": optax.sgd(learning_rate, momentum), "zero": optax.set_to_zero()}, 
                             freeze({"Dense_0": "train", "Conv_0": "zero"}))
  return train_state.TrainState.create(
      apply_fn=module.apply, params=params, tx=tx)


def train_step(state, batch):
  def loss_fn(params):
    logits = state.apply_fn({'params': params}, batch['image'])
    loss = optax.softmax_cross_entropy_with_integer_labels(
        logits=logits, labels=batch['label']).mean()
    return loss
  grad_fn = jax.grad(loss_fn)
  grads = grad_fn(state.params)
  state = state.apply_gradients(grads=grads)
  return state, grads


def get_datasets(num_epochs, batch_size):
  train_ds = tfds.load('mnist', split='train')
  test_ds = tfds.load('mnist', split='test')

  train_ds = train_ds.map(lambda sample: {'image': tf.cast(sample['image'],
                                                           tf.float32) / 255.,
                                          'label': sample['label']}) 
  test_ds = test_ds.map(lambda sample: {'image': tf.cast(sample['image'],
                                                         tf.float32) / 255.,
                                        'label': sample['label']}) 

  train_ds = train_ds.repeat(num_epochs).shuffle(1024)
  train_ds = train_ds.batch(batch_size, drop_remainder=True).prefetch(1)
  test_ds = test_ds.shuffle(1024)
  test_ds = test_ds.batch(batch_size, drop_remainder=True).prefetch(1) 

  return train_ds, test_ds


class CNN(nn.Module):
  @nn.compact
  def __call__(self, x):
    x = nn.Conv(features=32, kernel_size=(3, 3))(x)
    x = nn.relu(x)
    x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
    x = x.reshape((x.shape[0], -1))
    x = nn.Dense(features=10)(x)
    return x

num_epochs = 1
batch_size = 32
learning_rate = 1
momentum = 0.9
train_ds, test_ds = get_datasets(num_epochs, batch_size)
tf.random.set_seed(0)
num_steps_per_epoch = train_ds.cardinality().numpy() // num_epochs


cnn = CNN()
init_rng = jax.random.PRNGKey(0)
state = create_train_state(cnn, init_rng, learning_rate, momentum)
del init_rng  # Must not be used anymore.

for step,batch in enumerate(train_ds.as_numpy_iterator()):
  state, grads = train_step(state, batch)
  break  # one training iteration is enough

print(grads) # <- should be zero for Conv_0, it is not
```

**Comment by mkotyushev** [comment]
Hello! I also observe similar behavior: with usage of `multi_transform` + `set_to_zero` state (as it expected to be) is not changed after `apply_gradients`, but gradients are still calculated despite `jax.jit` usage on loss calculation function. 

In my understanding, gradients should not be calculated for params with `set_to_zero` transform (as it was pointed out by @mkunesch in #290 in the first comment). Am I missing something here?

Here is an example code to reproduce the behavior: state of `kernel2` is not changing, but its gradients are non-zero

```python
import jax
import optax
from typing import Any, Callable, Sequence
from jax import lax, random, numpy as jnp
from flax.core import freeze, unfreeze
from flax import linen as nn
from flax.training import train_state  

class SimpleDense(nn.Module):
  features: int
  kernel_init: Callable = nn.initializers.lecun_normal()
  bias_init: Callable = nn.initializers.zeros_init()

  @nn.compact
  def __call__(self, inputs):
    kernel1 = self.param('kernel1',
                        self.kernel_init, # Initialization function
                        (inputs.shape[-1], self.features // 2))  # shape info.
    kernel2 = self.param('kernel2',
                        self.kernel_init, # Initialization function
                        (inputs.shape[-1], self.features // 2))  # shape info.
    kernel = jnp.concatenate((kernel1, kernel2), axis=1)
    y = lax.dot_general(inputs, kernel,
                        (((inputs.ndim - 1,), (0,)), ((), ())),)
    bias = self.param('bias', self.bias_init, (self.features,))
    y = y + bias
    return y

@jax.jit
def train_step(state, x):
  def loss_fn(params):
    y = state.apply_fn({'params': params}, x)
    loss = ((x - y) ** 2).sum()
    return loss
  grad_fn = jax.grad(loss_fn)
  grads = grad_fn(state.params)
  state = state.apply_gradients(grads=grads)
  return state, grads

model = SimpleDense(features=4)

key1, key2 = random.split(random.PRNGKey(0), 2)
x = random.uniform(key1, (4,4))

params = model.init(random.PRNGKey(0), x)['params']
tx = optax.multi_transform({"train": optax.sgd(0.01), "zero": optax.set_to_zero()}, 
                            freeze({"kernel1": "train", "bias": "train", "kernel2": "zero"}))
opt_state = tx.init(params)

state = train_state.TrainState.create(
      apply_fn=model.apply, params=params, tx=tx)

for step in range(5):
  state, grads = train_step(state, x)
  print('kernel1')
  print('state: ', state.params['kernel1'][0])
  print('grads: ', grads['kernel1'][0])
  print('kernel2')
  print('state: ', state.params['kernel2'][0])
  print('grads: ', grads['kernel2'][0])
  print()```

**Comment by davisyoshida** [comment]
For anyone who comes across this in the future, the issue is that `grads` is being returned from `train_step`. The way the `set_to_zero()` strategy works is that it makes it so that the gradients for the frozen parameters are unused, allowing the JIT to remove the computation as dead code. It's not actually modifying the gradient computation, so if you return the grads that code is no longer dead and the JIT won't prune it.
